### PR TITLE
fix: Refactor permissions

### LINKED ISSUE #7049: Request and Revoke Permissions doesn't work properly
Problem with Deno and permissions.
When read access is requested for a specific folder, each read access request is either granted or denied, depending on what the user decides on the first request.

`deno run --unstable app.ts`
```js
// --> Deno requests read access to "./file1" --> grant
const status1 = await Deno.permissions.request({ name: "read", path: "./file1" });  
const status2 = await Deno.permissions.request({ name: "read", path: "./file2" });  // --> Deno skip request
console.log("Request ./file1", status1.state);    // --> Granted     | Denied 
console.log("Request ./file2", status2.state);    // --> Granted [!] | Denied [!]
```

Same for revoke permissions
`deno run --unstable --allow-read=./file1,./file2 app.ts`
case 1:
```js
// --> Deno revoke read access to "./file1"
const status1 = await Deno.permissions.revoke({ name: "read", path: "./file1" }); 
const status2 = await Deno.permissions.query({ name: "read", path: "./file1" });
const status3 = await Deno.permissions.query({ name: "read", path: "./file2" });
console.log("Request ./file1", status1.state);    // --> Granted [!]
console.log("Request ./file1", status2.state);    // --> Granted [!]
console.log("Request ./file2", status2.state);    // --> Granted
```

case 2:
```js
const status1 = await Deno.permissions.revoke({ name: "read" });  // --> Deno revoke read access
const status2 = await Deno.permissions.query({ name: "read", path: "./file1" });
const status3 = await Deno.permissions.query({ name: "read", path: "./file2" });
console.log("Request ./file1", status1.state);    // --> Prompt
console.log("Request ./file1", status2.state);    // --> Granted [!]
console.log("Request ./file2", status2.state);    // --> Granted [!]
```

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fix granular permission requests/revokes. Properly model the "denied" state.

Fixes #7049.

Internal structure before:

```rust
pub struct Permissions {
  pub allow_read: PermissionState,
  pub read_allowlist: HashSet<PathBuf>,
  pub allow_write: PermissionState,
  pub write_allowlist: HashSet<PathBuf>,
  pub allow_net: PermissionState,
  pub net_allowlist: HashSet<String>,
  pub allow_env: PermissionState,
  pub allow_run: PermissionState,
  pub allow_plugin: PermissionState,
  pub allow_hrtime: PermissionState,
}
```

Internal structure after:

```rust
pub struct UnaryPermission<T: Eq + Hash> {
  global_state: PermissionState,
  granted_list: HashSet<T>,
  denied_list: HashSet<T>,
}

pub struct Permissions {
  pub read: UnaryPermission<PathBuf>,
  pub write: UnaryPermission<PathBuf>,
  pub net: UnaryPermission<String>,
  pub env: PermissionState,
  pub run: PermissionState,
  pub plugin: PermissionState,
  pub hrtime: PermissionState,
}
```

See `docs/runtime/permission_apis.md` for the corrected user-facing semantics.

### CODE REVIEW

**Comment by ry** [comment]
@nayeemrmn Nice. I think this looks ok. I don't see the test case presented in https://github.com/denoland/deno/issues/7049 here - or did I miss it? It would be good to capture that.

File: cli/tests/integration_tests.rs [hunk 1]
```diff
@@ -1603,6 +1597,30 @@ itest!(_060_deno_doc_displays_all_overloads_in_details_view {
...
+  input: Some("g\nd\n"),
+  output: "061_permissions_request.ts.out",
+});
+
+// TODO(nayeemrmn): Prompts don't work because stdin isn't interactive here.
+itest_ignore!(_062_permissions_request_global {
+  args: "run --unstable 062_permissions_request_global.ts",
+  input: Some("g\n"),
+  output: "062_permissions_request_global.ts.out",
+});
```
Comment (nayeemrmn) [inline cli/tests/integration_tests.rs:1612]: Looking into adding a `pty` input option to `itest!()`...

**Comment by nayeemrmn** [comment]
@ry Ready for review 😩

File: docs/runtime/permission_apis.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+```
+
+Other examples:
+
+```ts
+// Global write permission.
+const desc1 = { name: "write" };
+
+// Write permission to `$PWD/foo/bar`.
+const desc2 = { name: "write", "foo/bar": "foo/bar" };
```
Comment (ry) [inline docs/runtime/permission_apis.md:29]: typo here. should be "path"

File: cli/tests/integration_tests.rs [hunk 3]
```diff
@@ -1603,6 +1597,70 @@ itest!(_060_deno_doc_displays_all_overloads_in_details_view {
   output: "060_deno_doc_displays_all_overloads_in_details_view.ts.out",
 });
 
+/// Kind of reflects `itest!()`. Note that the pty's output (which also contains
+/// stdin content) is compared against the content of the `output` path.
+#[cfg(not(windows))]
+fn itest_pty(args: &str, output: &str, input: &[u8]) {
```
Comment (ry) [inline cli/tests/integration_tests.rs:1603]: Can you move this function to the bottom of this file inside the "util" module?

Comment (nayeemrmn) [reply cli/tests/integration_tests.rs]: I've moved to the `test_util` crate, which I assume you meant.

Comment (ry) [reply cli/tests/integration_tests.rs]: Thank you, yes.

**Pull request merged**